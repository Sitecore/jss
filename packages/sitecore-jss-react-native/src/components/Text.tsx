import React from 'react';
import PropTypes from 'prop-types';
import { Text as NativeText } from 'react-native';

export interface TextProps {
  [nativeTextProps: string]: unknown;
  /** The text field data. */
  field: {
    value?: string;
    editable?: string;
  } | null;
}

export const Text: React.SFC<TextProps> = ({ field, ...otherProps }) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  const textValue = field.editable ? field.editable : field.value;

  return <NativeText {...otherProps}>{textValue}</NativeText>;
};

Text.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.any,
    editable: PropTypes.string,
  }),
};

Text.displayName = 'Text';
