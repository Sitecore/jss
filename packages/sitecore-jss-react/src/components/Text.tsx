import React, { ReactElement, FunctionComponent } from 'react';
import PropTypes from 'prop-types';

export interface TextProps {
  [htmlAttributes: string]: unknown;
  /** The text field data. */
  field?: {
    value?: string;
    editable?: string;
  };
  /**
   * The HTML element that will wrap the contents of the field.
   */
  tag?: string;
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   * @default true
   */
  editable?: boolean;
  /**
   * If false, HTML-encoding of the field value is disabled and the value is rendered as-is.
   */
  encode?: boolean;
}

export const Text: FunctionComponent<TextProps> = ({
  field,
  tag,
  editable,
  encode,
  ...otherProps
}) => {
  if (!field || (!field.editable && !field.value)) {
    return null;
  }

  // can't use editable value if we want to output unencoded
  if (!encode) {
    // eslint-disable-next-line no-param-reassign
    editable = false;
  }

  const value = (field.editable && editable ? field.editable : field.value) || '';

  let output: (ReactElement | string)[] = [value];

  // when value isn't formatted, we should format line breaks
  if (!field.editable && value) {
    const splitted = String(value).split('\n');

    if (splitted.length) {
      output = [];
    }

    splitted.forEach((str, i) => {
      const isLast = i === splitted.length - 1;

      output.push(str);

      if (!isLast) {
        output.push(<br key={i} />);
      }
    });
  }

  const setDangerously = (field.editable && editable) || !encode;

  let children = null;
  const htmlProps: {
    [htmlAttributes: string]: unknown;
    children?: React.ReactNode;
  } = {
    ...otherProps,
  };

  if (setDangerously) {
    htmlProps.dangerouslySetInnerHTML = {
      __html: output,
    };
  } else {
    children = output;
  }

  if (tag || setDangerously) {
    return React.createElement(tag || 'span', htmlProps, children);
  } else {
    return <React.Fragment>{children}</React.Fragment>;
  }
};

Text.propTypes = {
  field: PropTypes.shape({
    value: PropTypes.string,
    editable: PropTypes.string,
  }),
  tag: PropTypes.string,
  editable: PropTypes.bool,
  encode: PropTypes.bool,
};

Text.defaultProps = {
  editable: true,
  encode: true,
};

Text.displayName = 'Text';
