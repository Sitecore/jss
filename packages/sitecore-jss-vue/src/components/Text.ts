import { h, defineComponent } from 'vue';

export interface TextProps {
  /** The text field data. */
  field: {
    value?: string | number;
    editable?: string;
  };
  /**
   * The HTML element that will wrap the contents of the field.
   */
  tag?: string;
  /**
   * Can be used to explicitly disable inline editing.
   * If true and `field.editable` has a value, then `field.editable` will
   * be processed and rendered as component output. If false, `field.editable` value will be ignored and not rendered.
   */
  editable?: boolean;
  /**
   * If false, HTML-encoding of the field value is disabled and the value is rendered as-is.
   */
  encode?: boolean;
}

export const Text = defineComponent({
  props: {
    field: { type: Object, required: true },
    tag: { type: String, default: 'span' },
    editable: { type: Boolean, default: true },
    encode: { type: Boolean, default: true },
  },
  render() {
    const { field, tag, editable, encode } = this.$props;
    if (!field || (!field.editable && (field.value === undefined || field.value === ''))) {
      return null;
    }

    // can't use editable value if we want to output unencoded
    const isEditable = !encode ? false : editable;

    const output = field.editable && isEditable ? field.editable : field.value;
    const setDangerously = (field.editable && isEditable) || !encode;

    // this.$data should be passed along to the
    // `createElement` function in order to retain attributes and events
    // https://v3.vuejs.org/guide/render-function.html#render-functions
    const data: any = { ...this.$data };

    let children = null;
    if (setDangerously) {
      data.innerHTML = output;
    } else {
      children = output;
    }

    return h(tag || 'span', data, children);
  },
});
