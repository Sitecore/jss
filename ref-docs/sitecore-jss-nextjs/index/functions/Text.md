[**@sitecore-jss/sitecore-jss-nextjs**](../../README.md) â€¢ **Docs**

***

[@sitecore-jss/sitecore-jss-nextjs](../../README.md) / [index](../README.md) / Text

# Function: Text()

> **Text**(`props`, `deprecatedLegacyContext`?): `ReactNode`

## Parameters

| Parameter | Type | Description |
| ------ | ------ | ------ |
| `props` | `TextProps` | - |
| `deprecatedLegacyContext`? | `any` | **Deprecated** **See** [React Docs](https://legacy.reactjs.org/docs/legacy-context.html#referencing-context-in-lifecycle-methods) |

## Returns

`ReactNode`

## Defined in

sitecore-jss-react/types/components/Text.d.ts:22
