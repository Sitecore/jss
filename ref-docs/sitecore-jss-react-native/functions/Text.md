[**@sitecore-jss/sitecore-jss-react-native**](../README.md) • **Docs**

***

[@sitecore-jss/sitecore-jss-react-native](../README.md) / Text

# Function: Text()

> **Text**(`props`, `context`?): `ReactElement`\<`any`, `any`\>

## Parameters

• **props**: `PropsWithChildren`\<`TextProps`\>

• **context?**: `any`

## Returns

`ReactElement`\<`any`, `any`\>

## Defined in

[sitecore-jss-react-native/src/components/Text.tsx:14](https://github.com/Sitecore/jss/blob/ff400466a8d16483c667d9a837e1247d6192035e/packages/sitecore-jss-react-native/src/components/Text.tsx#L14)
