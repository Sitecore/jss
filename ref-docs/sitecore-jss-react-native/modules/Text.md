[@sitecore-jss/sitecore-jss-react-native](../README.md) / Text

# Namespace: Text

## Table of contents

### Variables

- [displayName](Text.md#displayname)
- [propTypes](Text.md#proptypes)

## Variables

### displayName

• **displayName**: `string`

#### Defined in

[sitecore-jss-react-native/src/components/Text.tsx:31](https://github.com/Sitecore/jss/blob/3ea859f70/packages/sitecore-jss-react-native/src/components/Text.tsx#L31)

___

### propTypes

• **propTypes**: `WeakValidationMap`\<`TextProps`\>

#### Defined in

[sitecore-jss-react-native/src/components/Text.tsx:24](https://github.com/Sitecore/jss/blob/3ea859f70/packages/sitecore-jss-react-native/src/components/Text.tsx#L24)
