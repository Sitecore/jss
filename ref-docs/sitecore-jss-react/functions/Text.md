[**@sitecore-jss/sitecore-jss-react**](../README.md) â€¢ **Docs**

***

[@sitecore-jss/sitecore-jss-react](../README.md) / Text

# Function: Text()

> **Text**(`props`, `deprecatedLegacyContext`?): `ReactNode`

## Parameters

| Parameter | Type | Description |
| ------ | ------ | ------ |
| `props` | `TextProps` | - |
| `deprecatedLegacyContext`? | `any` | **Deprecated** **See** [React Docs](https://legacy.reactjs.org/docs/legacy-context.html#referencing-context-in-lifecycle-methods) |

## Returns

`ReactNode`

## Defined in

[packages/sitecore-jss-react/src/components/Text.tsx:28](https://github.com/Sitecore/jss/blob/b543e221483be0d7e4e3ae7b76785619d291d2d3/packages/sitecore-jss-react/src/components/Text.tsx#L28)
