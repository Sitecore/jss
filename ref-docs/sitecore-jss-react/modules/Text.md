[@sitecore-jss/sitecore-jss-react](../README.md) / Text

# Namespace: Text

## Table of contents

### Variables

- [defaultProps](Text.md#defaultprops)
- [displayName](Text.md#displayname)
- [propTypes](Text.md#proptypes)

## Variables

### defaultProps

• **defaultProps**: `Partial`\<`TextProps`\>

#### Defined in

[sitecore-jss-react/src/components/Text.tsx:110](https://github.com/Sitecore/jss/blob/a9a4cb30d/packages/sitecore-jss-react/src/components/Text.tsx#L110)

___

### displayName

• **displayName**: `string`

#### Defined in

[sitecore-jss-react/src/components/Text.tsx:115](https://github.com/Sitecore/jss/blob/a9a4cb30d/packages/sitecore-jss-react/src/components/Text.tsx#L115)

___

### propTypes

• **propTypes**: `WeakValidationMap`\<`TextProps`\>

#### Defined in

[sitecore-jss-react/src/components/Text.tsx:100](https://github.com/Sitecore/jss/blob/a9a4cb30d/packages/sitecore-jss-react/src/components/Text.tsx#L100)
