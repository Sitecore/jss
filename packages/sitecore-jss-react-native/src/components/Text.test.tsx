import React from 'react';
import renderer from 'react-test-renderer';
import { Text } from './Text';

describe('<Text />', () => {
  test('should render nothing with missing field', () => {
    const field = null;
    const rendered = renderer.create(<Text field={field} />);
    expect(rendered).toMatchSnapshot();
  });

  test('should render nothing with missing editable and value', () => {
    const props = {
      field: {},
    };
    const rendered = renderer.create(<Text {...props} />);
    expect(rendered).toMatchSnapshot();
  });

  test('should render editable value if present', () => {
    const props = {
      field: {
        value: 'this value should not be rendered',
        editable: 'this value should be rendered',
      },
    };
    const rendered = renderer.create(<Text {...props} />);

    expect(rendered).toMatchSnapshot();
  });

  test('should render value if editable is undefined', () => {
    const props = {
      field: {
        value: 'this value should be rendered',
        editable: '',
      },
    };
    const rendered = renderer.create(<Text {...props} />);

    expect(rendered).toMatchSnapshot();
  });

  it('should render other attributes with other props provided', () => {
    const field = {
      editable: 'this value should be rendered',
    };

    const rendered = renderer.create(<Text field={field} style={{ flex: 1 }} selectable={true} />);
    expect(rendered).toMatchSnapshot();
  });
});
