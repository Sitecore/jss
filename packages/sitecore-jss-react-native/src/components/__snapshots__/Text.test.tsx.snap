// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> should render editable value if present 1`] = `
<Text>
  this value should be rendered
</Text>
`;

exports[`<Text /> should render nothing with missing editable and value 1`] = `null`;

exports[`<Text /> should render nothing with missing field 1`] = `null`;

exports[`<Text /> should render other attributes with other props provided 1`] = `
<Text
  selectable={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  this value should be rendered
</Text>
`;

exports[`<Text /> should render value if editable is undefined 1`] = `
<Text>
  this value should be rendered
</Text>
`;
